### ISSUE TITLE
go-plugin + static routes plugin

### ISSUE THREAD

**Issue by jonmorehouse** [jonmorehouse/gatekeeper#3]
So the next step after we've implemented some basic networking over in #2 is to set up some basic scaffolding for fetching the correct route information for a request.

Thinking through this now, it seems like we should "upstream" plugins which for now is just a plugin returning static urls. It seems to me like the `gatekeepers` primary process should maintain a local cache of the upstreams it knows about, whilst exposing a method for individual plugins to call into and update.

As a first pass, it can be single-direction communication and each request to the proxy can call into the `upstreams` plugin (no caching yet).

Off the top of my head, each upstream should have the following parameters;
- prefix
- hostnames 
- protocols (http-internal, http-public)

As a first pass, I think we can make the following assumption:

If _hostnames_ is passed in, any request with a `Host` header linked to the upstream will _NOT_ strip the prefix. _eg:_

``` bash
curl -H 'Host: my-upstream' gatekeeper/foo/bar => foo_upstream/foo/bar
```

If a request just comes into the gatekeeper api with just a prefix, then we can assume that the prefix should be stripped before proxying on to the upstream.

``` bash
curl gatekeeper/my-upstream/foo/bar => foo_upstream/bar
```

**Comment by jonmorehouse** [comment]
Just thinking through some directory structure here:

```
tree .
.
├── bins
│   ├── gatekeeper
│   ├── plugin-consul-upstreams
│   └── plugin-static-upstreams
├── gatekeeper
│   └── lib.go
├── main.go
└── plugins
    ├── plugin-consul-upstreams
    └── plugin-static-upstreams
```

Here's how I think this project should come out structured...

Notably, different "builtin" plugins are contained in the `plugins` directory. `gatekeeper` will be a module which is shared by the top level `main` package and modules within the `plugins` directory. In order to build any particular package, we'll simply need to add the gatekeeper module to our GOPATH in a deterministic manner.

**Comment by jonmorehouse** [comment]
So coming up with a parent-child plugin relationship ended up being tricky; here's some example code showing how this could all come out...

https://gist.github.com/jonmorehouse/75e02047ec5ac47a8109ba0ba13eea80

**Comment by jonmorehouse** [comment]
Closed by #7
